File: Hot 100/Backtracing/0017_LetterCombinationsOfAPhoneNumber.java
class Solution {
    private static final String[] arr = {
        "",
        "",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };

    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.isEmpty()) {
            return new ArrayList<>();
        }

        List<String> ans = new ArrayList<>();

        backtrack(digits, 0, new StringBuilder(), ans);

        return ans;
    }

    private void backtrack(String digits, int idx, StringBuilder sb, List<String> ans) {
        if (idx == digits.length()) {
            ans.add(sb.toString());
            return;
        }

        String str = arr[digits.charAt(idx) - '0'];

        for (char c : str.toCharArray()) {
            sb.append(c);
            backtrack(digits, idx + 1, sb, ans);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}

File: Top 100 Liked/Backtracking/0017_letter_combinations_of_a_phone_number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        phone = {
            '2': "abc", '3': "def", '4': "ghi", '5': "jkl",
            '6': "mno", '7': "pqrs", '8': "tuv", '9': "wxyz"
        }
        ans = []

        def backtrack(idx: int, path: List[str]):
            if idx == len(digits):
                ans.append(''.join(path))
                return

            combo = phone[digits[idx]]

            for ch in combo:
                path.append(ch)
                backtrack(idx + 1, path)
                path.pop()

        backtrack(0, [])
        return ans
